HARK/ConsumptionSaving/ConsLabeledModel.py

```diff
@@ -0,0 +1,663 @@
+        )
+        result.attrs = self.dataset["v"].attrs
+
+        return result
+
+
+class ConsumerSolutionLabeled(MetricObject):
+    def __init__(
+        self,
+        value: ValueFuncCRRALabeled,
+        policy: xr.Dataset,
+        continuation: ValueFuncCRRALabeled,
+        attrs=None,
+    ):
+
+        if attrs is None:
+            attrs = dict()
+
+        self.value = value  # value function
+        self.policy = policy  # policy function
+        self.continuation = continuation  # continuation function
+
+        self.attrs = attrs
+
+    def distance(self, other):
+
+        value = self.value.dataset
+        other_value = other.value.dataset.interp_like(value)
+
+        return np.max(np.abs(value - other_value).to_array())
+
+
+class PerfForesightLabeledType(IndShockConsumerType):
+    def __init__(self, verbose=1, quiet=False, **kwds):
+        params = init_perfect_foresight.copy()
+        params.update(kwds)
+
+        # Initialize a basic AgentType
+        IndShockConsumerType.__init__(self, verbose=verbose, quiet=quiet, **params)
+
+        self.solve_one_period = make_one_period_oo_solver(
+            ConsPerfForesightLabeledSolver
+        )
+
+    def update_solution_terminal(self):
+
+        u = lambda c: CRRAutility(c, self.CRRA)  # utility
+        uP = lambda c: CRRAutilityP(c, self.CRRA)  # marginal utility
+
+        mNrm = xr.DataArray(
+            np.append(0.0, self.aXtraGrid),
+            name="mNrm",
+            dims=("mNrm"),
+            attrs={"long_name": "cash_on_hand"},
+        )
+        state = xr.Dataset({"mNrm": mNrm})  # only one state var in this model
+
+        # optimal decision is to consume everything in the last period
+        cNrm = xr.DataArray(
+            mNrm,
+            name="cNrm",
+            dims=state.dims,
+            coords=state.coords,
+            attrs={"long_name": "consumption"},
+        )
+
+        v = u(cNrm)
+        v.name = "v"
+        v.attrs = {"long_name": "value function"}
+
+        v_der = uP(cNrm)
+        v_der.name = "v_der"
+        v_der.attrs = {"long_name": "marginal value function"}
+
+        v_inv = cNrm.copy()
+        v_inv.name = "v_inv"
+        v_inv.attrs = {"long_name": "inverse value function"}
+
+        v_der_inv = cNrm.copy()
+        v_der_inv.name = "v_der_inv"
+        v_der_inv.attrs = {"long_name": "inverse marginal value function"}
+
+        dataset = xr.Dataset(
+            {
+                "cNrm": cNrm,
+                "v": v,
+                "v_der": v_der,
+                "v_inv": v_inv,
+                "v_der_inv": v_der_inv,
+            }
+        )
+
+        vfunc = ValueFuncCRRALabeled(
+            dataset[["v", "v_der", "v_inv", "v_der_inv"]], self.CRRA
+        )
+
+        self.solution_terminal = ConsumerSolutionLabeled(
+            value=vfunc,
+            policy=dataset[["cNrm"]],
+            continuation=None,
+            attrs={"m_nrm_min": 0.0},
```

Comment:
whereas `value`, `policy`, and `continuation` are all objects at the level of abstraction of the Bellman problem, `attrs` is as the level of abstraction of the underlying xarray-powered data structure.

That's somewhat mismatched, and I wonder about the meaning of the `"m_nrm_min" field here. Is this representing a constraint on the policy, or a description of the state grid?

(Also, we should probably use _either_ `cNrm` or `c_nrm` but not switch between them?)